### ISSUE TITLE
Suggestion: prettierLichess body class

### ISSUE THREAD

**Issue by Siderite** [prettierlichess/prettierlichess#168]
I am trying to style my own extension and I have to take into consideration the possibility that the user has Prettier installed. However, creating rules that fit all possibilities at once is not easy. Adding a global body class for when prettier is installed (and maybe make all CSS classes consider it rather than putting !important everywhere :) ) would help others tailor their work to handle this.

**Comment by kieferro** [comment]
I can understand that it is difficult when multiple extensions collide. I can gladly change the extension so that a class is added to the `body` tag when the extension is loaded. Another option would be to have a meta tag with a special name, so that you can easily check if the extension is installed:
```js
if (document.querySelector('meta[name="prettierlichess-loaded"]'))
```
Theoretically, I can also do both. Just let me know what makes the most sense to you (PRs are also welcome if you would be interested).



> rather than putting !important everywhere

I am very much aware that using `!important` is not a good solution if someone else wants to overwrite the style. The problem is just unfortunately that CSS rules without `!important` only override other rules if they are more specific. But we can't guarantee that for every selector. For that we would have to check which selectors lichess uses and then construct more specific ones (which sometimes doesn't really make sense, because the possibility of the extension to really override the styling for all elements is just based on the fact that there are many general rules).
So I understand this point well, but I don't really see a solution to it. As far as I know, CSS just doesn't provide a good solution for this (like `!important` with multiple priorities) or am I misinformed?

**Comment by Siderite** [comment]
I understand your points, I really do, but at the same time you have to realize there are some situations where it's almost impossible to change the style once !important is involved. For example keyframe animations. CSS does not allow important in them and if the current style has !important I cannot override it at all.

The reason why I recommended a global CSS class is that now any selector can be easily be made more selective, you can detect if Prettier is active and you can also disable the styles by just removing the class, which means you can check the original styles without having to go to Extensions, disabling Prettier, then refreshing the page. So many reasons come together.

And then we could ask the same thing to the LiChess devs, in which case removing the theme CSS they would use would remove all styles from the website, eliminating the need for !important as well.

I guess I am thinking in namespaces, I am a .NET dev, can't help it :D

Now, as the title says, it's a suggestion. And a meaney hint on how I had to add !important to my CSS because Prettier did it first... 👿

**Comment by kieferro** [comment]
> I recommended a global CSS class

I don't think you quite understand the extent of what this would entail. This would require some simple changes: adding all the selectors, rewriting the script so that it inserts the CSS class. You would also have to make sure that it happens fast enough to not have a sudden flash effect. That would require a bit of work, but still wouldn't be too big of a problem.
But then you would have to go through all the rules again (including the different `@media`-queries) and check for which ones the specificity is not high enough by looking at the website. Then you would have to find a more specific selector for them which still selects the same group of elements. You would certainly miss a few elements and it would mean that the extension would not be able to react as well on its own to newly added elements without making changes.

That is an amount of problems and work that just isn't worth it.


> And then we could ask the same thing to the LiChess devs, in which case removing the theme CSS they would use would remove all styles from the website, eliminating the need for !important as well.

Feel free to do so, I'd actually be interested in what they have to say about that.






> I guess I am thinking in namespaces, I am a .NET dev, can't help it :D



I also think it's stupid that CSS doesn't have a namespace solution or `!important` with different priorities. And believe me, not being able to use `!important` for keyframe animations has caused me some problems too.
It would all be nicer if CSS was structured differently, but that's just not the way it is and if that would change at some point, I'm extremely grateful to the people who contributed to it and I'm happy to include any new solutions to help with the problems you describe. But as long as that is not the case, I will not risk that the styling for many people will be significantly worse and that extremely much work must be invested just to support this.
